src/ray/gcs/subscription_executor.cc

```diff
@@ -0,0 +1,131 @@
+#include "ray/gcs/subscription_executor.h"
+
+namespace ray {
+
+namespace gcs {
+
+template <typename ID, typename Data, typename Table>
+Status SubscriptionExecutor<ID, Data, Table>::AsyncSubscribe(
+    const ClientID &client_id, const SubscribeCallback<ID, Data> &subscribe,
+    const StatusCallback &done) {
+  // TODO(micafan) Optimize the lock when necessary.
+  // Consider avoiding locking in single-threaded processes.
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (subscribe != nullptr && subscribe_all_callback_ != nullptr) {
+    RAY_LOG(INFO) << "Duplicate subscription!";
+    return Status::Invalid("Duplicate subscription!");
+  }
+
+  if (registered_) {
+    return Status::OK();
+  }
+
+  auto on_subscribe = [this](RedisGcsClient *client, const ID &id,
+                             const std::vector<Data> &result) {
+    if (result.empty()) {
+      return;
+    }
+
+    SubscribeCallback<ID, Data> sub_one_callback = nullptr;
+    SubscribeCallback<ID, Data> sub_all_callback = nullptr;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      const auto it = id_to_callback_map_.find(id);
+      if (it != id_to_callback_map_.end()) {
+        sub_one_callback = it->second;
+      }
+      sub_all_callback = subscribe_all_callback_;
+    }
+    if (sub_one_callback != nullptr) {
+      sub_one_callback(id, result.back());
+    }
+    if (sub_all_callback != nullptr) {
+      sub_all_callback(id, result.back());
+    }
+  };
+
+  auto on_done = [this, done](RedisGcsClient *client) {
+    if (done != nullptr) {
+      done(Status::OK());
+    }
+  };
+
+  Status status = table_.Subscribe(JobID::Nil(), client_id, on_subscribe, on_done);
+  if (status.ok()) {
+    registered_ = true;
+    subscribe_all_callback_ = subscribe;
+  }
+
+  return status;
+}
+
+template <typename ID, typename Data, typename Table>
+Status SubscriptionExecutor<ID, Data, Table>::AsyncSubscribe(
+    const ClientID &client_id, const ID &id, const SubscribeCallback<ID, Data> &subscribe,
+    const StatusCallback &done) {
+  Status status = AsyncSubscribe(client_id, nullptr, nullptr);
+  if (!status.ok()) {
+    return status;
+  }
+
+  auto on_done = [this, done, id](Status status) {
+    if (!status.ok()) {
+      std::lock_guard<std::mutex> lock(mutex_);
+      id_to_callback_map_.erase(id);
+    }
+    if (done != nullptr) {
+      done(status);
+    }
+  };
+
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    const auto it = id_to_callback_map_.find(id);
+    if (it != id_to_callback_map_.end()) {
+      RAY_LOG(INFO) << "Duplicate subscription to id " << id << " client_id "
+                    << client_id;
+      return Status::Invalid("Duplicate subscription!");
+    }
+    status = table_.RequestNotifications(JobID::Nil(), id, client_id, on_done);
```

Comment:
I got a bit confused on the behavior here. In the caller of AsyncSubscribe(), it's passed a ClientID::Nil() so that it would subscribe the whole table even when user subscribes only one actor, and in that case the callbacks added to `id_to_callback_map_` will always get invoked, then we don't need the `RequestNotifications` above?

Reply:
Fixing...